main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
+      case Binary(_, _, _, _, _) => None
+      case IfThenElse(_, _, _, _, _) => None
+      case Let(_, _, _, _, _) => None
+      case LetRec(_, _, _, _, _) => None
+      case Is(_, _, _, _) => None
+      case Tag(_, _, _, _, _) => None
+      case Untag(_, _, _, _, _) => None
+      case Index(_, _, _, _) => None
+      case Tuple(_, _, _) => None
+      case Existential(_, _, _) => None
+      case Universal(_, _, _) => None
+      case NativeConstructor(_, _, _, _) => None
+      case NativeField(_, _, _) => None
+      case NativeMethod(_, _, _, _) => None
+      case UserError(_, _) => None
+      case MatchError(_, _) => None
+      case SwitchError(_, _) => None
+      /* Error */
+      case MkClosure(_, _, _, _) =>
+        throw InternalCompilerException(s"Unexpected expression $exp after lambda lifting")
+    }
+  }
+
+  /**
+    * Traverses through `exp` and performs inlining using
+    * definitions and sizes from `scores`
+    */
+  def inlineExpr(definitions: Map[DefnSym, Constant], scores: Map[DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    def f = inlineVisitor(definitions,scores,_: Expression)
+    AstVisitor.visitExpressionWith(f, exp)
   }
 
+}
+
+object FreshenExpr {
+
+  /**
+    * Freshen all let binds in an expression
+    */
+  def freshenExpr(exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    val (m,newLets) = freshSymbols(exp)
+    Inliner.substituteWith(m,newLets)
+  }
+
+  def freshSymbols(exp:Expression)(implicit genSym: GenSym): (Map[VarSym, VarSym], Expression) =
+  {
+    var m = Map.empty[VarSym, VarSym]
+
+    def freshExprVisitor(exp3:Expression): Option[Expression] =
+    {
+      exp3 match {
+        case Let(sym, exp1, exp2, tpe, loc) =>
+          val newSym = Symbol.freshVarSym("inlnrFresh")
+          m += (sym -> newSym)
+          val freshExp1 = AstVisitor.visitExpressionWith(freshExprVisitor, exp1)
+          val freshExp2 = AstVisitor.visitExpressionWith(freshExprVisitor, exp2)
+          Some(Let(newSym,freshExp1,freshExp2,tpe,loc))
+        case LetRec(sym, exp1, exp2, tpe, loc) =>
+          val newSym = Symbol.freshVarSym("inlnrFresh")
+          m += (sym -> newSym)
+          val freshExp1 = AstVisitor.visitExpressionWith(freshExprVisitor, exp1)
+          val freshExp2 = AstVisitor.visitExpressionWith(freshExprVisitor, exp2)
+          Some(LetRec(newSym,freshExp1,freshExp2,tpe,loc))
+        case _ => None
+      }
+    }
+
+    val newLets = AstVisitor.visitExpressionWith(freshExprVisitor, exp)
+    (m,newLets)
+  }
+
+}
+
+object AstVisitor {
+
+  /**
+    * Expression Visitor function for SimplifiedAST
+    *
+    * @param f function that takes in a Expression and returns a Option[Expression]
+    * @param exp Expression to visit
+    * @return if f(exp) is Some(e) then e, else recursively visit expressions in exp
+    */
+  def visitExpressionWith(f: Expression => Option[Expression], exp: Expression): Expression =
```

Comment:
As you can see here, changing the AST involves upwards 20 files:
https://github.com/flix/flix/pull/474/files